### ISSUE TITLE
BSR SpMV: illegal scratchpad accesses with cuda, float

### ISSUE THREAD

**Issue by brian-kelley** [kokkos/kokkos-kernels#1625]
In the BSR SpMV test ``cuda.sparse_bsrmat_times_vec_float_int_size_t_TestExecSpace``, there is some kind of scratchpad (shared memory) overrun happening. On my workstation GPU (arch Pascal60), in a debug build, Kokkos prints out a bunch of messages like this while the kernel's running:
```
ScratchMemorySpace<...>::get_shmem: Failed to allocate 36 byte(s); remaining capacity is 36 byte(s)
```
which is a bit weird because it looks like there's exactly enough scratch available.

I think this rules out the tensor core path for this issue since Pascal doesn't have that hardware.

On weaver (I guess without Kokkos debug enabled), I just get:
```
[ RUN      ] cuda.sparse_bsrmat_times_vec_float_int_size_t_TestExecSpace
cudaDeviceSynchronize() error( cudaErrorIllegalAddress): an illegal memory access was encountered /ascldap/users/bmkelle/weaverMisc/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:133
```

I didn't see anything obviously wrong in the code. I was thinking maybe the test used mixed scalar types and then sized scratch for the wrong type, but that doesn't appear to be the case (matrix's values, x and y all have the same value type).

Replicating on Weaver Rhel8 node, usnig develop branch kokkos, kernels from today:
```
source /etc/profile.d/modules.sh
module purge
module load git cmake/3.21.2 cuda/11.2.2/gcc/8.3.1 openmpi/4.1.1/gcc/8.3.1/cuda/11.2.2 openblas/0.3.18/gcc/8.3.1
```
```
../kokkos-kernels/cm_generate_makefile.bash \
  --with-devices=Cuda \
  --arch=Power9,Volta70 \
  --cxxstandard="17" \
  --with-scalars=float
```

Most of this was written by the NGA folks, so I think @cwpearson is the only one we have left who is decently familiar with BSR SPMV.

**Comment by cwpearson** [comment]
I’ll try to figure this out

**Comment by brian-kelley** [comment]
Maybe related? https://github.com/kokkos/kokkos/issues/1811
Kokkos core bug involving shared memory size calcuation, although it seems to involve scratch views and BSR spmv here allocates raw shared memory.

**Comment by PhilMiller** [comment]
The fix for kokkos/kokkos#1811 was just merged to `develop`. Please give that a try.

**Comment by brian-kelley** [comment]
I'm looking through old issues, and just tried both of the original reproducers. Neither of them have issues now. Thanks @PhilMiller !
